src/suggestion.c

```diff
@@ -0,0 +1,399 @@
+    new_prefix[len] = new_prefix[len - 1];
+    new_prefix[len - 1] = suffix[0];
+    new_prefix[len + 1] = '\0';
+
+    if (has_children(t, new_prefix) == EXIT_SUCCESS) {
+        // Adding 1 to the suffix pointer will essentially delete the first character
+        rc += suggestions(set, t, new_prefix, suffix + 1, edits_left - 1, max_matches);
+    }
+
+    // Save some space now that we're done
+    free(new_prefix);
+
+    return rc + EXIT_SUCCESS;
+}
+
+// Helper function for suggestions that tries to insert a character to the end of a prefix
+int try_insert(match_t **set, trie_t *t, char *prefix, char *suffix, int edits_left, int max_matches) {
+
+    int i;
+    int rc = 0;
+    int len = strlen(prefix);
+    char* new_prefix;
+
+    // Ran out of space
+    if (len == MAXLEN - 1) {
+        return rc + EXIT_FAILURE;
+    }
+
+    for (i = 1; i < 248; i++) {
+
+        char c = (char)i;
+
+        if (trie_char_exists(t, c) == EXIT_SUCCESS) {
+            
+            new_prefix = malloc(sizeof(char) * (MAXLEN + 1));
+            if (new_prefix == NULL) {
+                return rc + EXIT_FAILURE;
+            } else {
+
+                strncpy(new_prefix, prefix, MAXLEN);
+
+                // Try adding on a new character
+                new_prefix[len] = c;
+                new_prefix[len + 1] = '\0';
+
+                if (has_children(t, new_prefix) == EXIT_SUCCESS) {
+
+                    // Basically just inserting the new ASCII character to the string
+                    rc += suggestions(set, t, new_prefix, suffix, edits_left - 1, max_matches);
+                } 
+            
+                // Save some space now that we're done
+                free(new_prefix);
+            }
+        }
+    }
+
+    return rc + EXIT_SUCCESS;
+}
+
+// Helper function for suggestions(). Attempts to add a match to a suggestion set
+int attempt_add(match_t **set, trie_t *t, char *s, int edits_left, int max_matches) {
+    int i;
+
+    // Check if the current string is in the trie
+    if (trie_search(t, s) == 1) {
+
+        // Look for the string in the set to update it, or add it
+        for (i = 0; i < max_matches; i++) {
+
+            if (set[i] == NULL) {
+
+                // String does not exist in the set, so add it
+
+                set[i] = (match_t*)malloc(sizeof(match_t));
+                if (set[i] == NULL) {
+                    return EXIT_FAILURE;
+                }
+
+                set[i]->str = malloc(sizeof(char) * MAXLEN);
+                if (set[i]->str == NULL) {
+                    return EXIT_FAILURE;
+                }
+                strcpy(set[i]->str, s);
+                set[i]->edits_left = edits_left;
+
+                if (set[i]->str == NULL) {
+                    // strcpy failed
+                    return EXIT_FAILURE;
+                }
+
+                // We don't want to trigger adding in place of the worst match
+                i = max_matches;
```

Comment:
What's the purpose of this line, considering that you're immediately returning after this line?

Reply:
@dwahme

Reply:
Good catch, I was trying to mirror the section below but this is unnecessary. Removed.